File: swift/Tests/UInt256Tests.swift
// Copyright © 2017-2019 Trust.
//
// This file is part of Trust. The full Trust copyright notice, including
// terms governing use, modification, and redistribution, is contained in the
// file LICENSE at the root of the source code distribution tree.

import TrustWalletCore
import XCTest

class UInt256Tests: XCTestCase {
    func testCreateWithData() {
        let data = Data(bytes: Array(repeating: 0, count: 24) + [0x1b, 0xc1, 0x6d, 0x67, 0x4e, 0xc8, 0x00, 0x00])
        let number = UInt256(data: data)!

        XCTAssertEqual(number.format(decimals: 18, exponent: 0), "2.0")
        XCTAssertEqual(Array(number.data), Array(data))
    }

    func testZeroData() {
        XCTAssertEqual(Array(UInt256.zero.data), Array(repeating: 0, count: 32))
    }

    func testIsZero() {
        XCTAssertTrue(UInt256.zero.isZero)
        XCTAssertFalse(UInt256.one.isZero)
    }

    func testFormat() {
        XCTAssertEqual(UInt256.zero.format(decimals: 0, exponent: 0), "0")
        XCTAssertEqual(UInt256.one.format(decimals: 0, exponent: 0), "1")
        XCTAssertEqual(UInt256.zero.format(decimals: 1, exponent: 1), "0.0")

        let x = UInt256(1234567890 as UInt64)
        XCTAssertEqual(x.format(decimals: 0, exponent: 0), "1234567890")
        XCTAssertEqual(x.format(decimals: 1, exponent: 1), "1234567890.0")
        XCTAssertEqual(x.format(decimals: 5, exponent: 0), "12345.6789")
        XCTAssertEqual(x.format(decimals: 5, exponent: 1), "123456.789")
    }
}

File: tests/Uint256Tests.cpp
// Copyright © 2017-2019 Trust.
//
// This file is part of Trust. The full Trust copyright notice, including
// terms governing use, modification, and redistribution, is contained in the
// file LICENSE at the root of the source code distribution tree.

#include "gtest/gtest.h"
#include "../src/UInt256.h"

using namespace TW;

TEST(UInt256Tests, ZeroIsZero) {
    auto zero = UInt256();
    for (auto i = 0; i < UInt256::wordCount; i+= 1) {
        ASSERT_EQ(zero.words[i], 0);
    }
}

TEST(UInt256Tests, Add) {
    auto zero = UInt256();
    auto num = UInt256(12345u);
    ASSERT_EQ(zero + num, num);
}

TEST(UInt256Tests, AddOverflow) {
    auto number = UInt256();
    auto expected = UInt256();
    for (auto i = 0; i < UInt256::wordCount; i+= 1) {
        number.words[i] = 0xffff'ffff'ffff'ffff;
    }
    number.words[UInt256::wordCount - 1] = 0x7fff'ffff'ffff'ffff;
    expected.words[UInt256::wordCount - 1] = 0x8000'0000'0000'0000;
    auto result = number + UInt256(1u);
    
    ASSERT_EQ(result, expected);
}

TEST(UInt256Tests, Multiply) {
    auto number = UInt256();
    auto expected = UInt256();
    for (auto i = 0; i < UInt256::wordCount; i+= 1) {
        number.words[i] = 0xffff'ffff'ffff'ffff;
        expected.words[i] = 0xffff'ffff'ffff'ffff;
    }
    number.words[UInt256::wordCount - 1] = 0x0fff'ffff'ffff'ffff;
    expected.words[0] = 0xffff'ffff'ffff'fffe;
    expected.words[UInt256::wordCount - 1] = 0x1fff'ffff'ffff'ffff;
    auto result = number * UInt256(2u);

    ASSERT_EQ(result, expected);
}
